packages/flutter/lib/src/material/carousel.dart

```diff
@@ -1211,6 +1344,42 @@ class _RenderSliverWeightedCarousel extends RenderSliverFixedExtentBoxAdaptor {
         ? getMaxChildIndexForScrollOffset(targetEndScrollOffsetForPaint, deprecatedExtraItemExtent)
         : null;
 
+    // Calculate the closest item to the center of the viewport.
+    // This is used to determine the current item index for the carousel.
+    final double center = constraints.scrollOffset + constraints.viewportMainAxisExtent / 2;
+
+    RenderBox? currentChild = firstChild;
+    int? newIndex;
+    double minDistance = double.infinity;
+
+    while (currentChild != null) {
+      final SliverMultiBoxAdaptorParentData parentData =
+          currentChild.parentData! as SliverMultiBoxAdaptorParentData;
+      final int index = parentData.index!;
+      final double layoutOffset = parentData.layoutOffset ?? 0.0;
+      final double itemExtent = _buildItemExtent(index, _currentLayoutDimensions);
+      final double item = layoutOffset + itemExtent / 2;
+
+      final double distance = (center - item).abs();
+      if (distance < minDistance) {
+        minDistance = distance;
+        newIndex = index;
+      }
+
+      currentChild = childAfter(currentChild);
+    }
+
+    if (newIndex != null && _currentIndex != newIndex) {
+      _currentIndex = newIndex;
+      controller._currentIndex = newIndex;
+
+      if (_hasDispatchedIndexOnce) {
+        onIndexChanged?.call(newIndex);
```

Comment:
The `performLayout` method is used to determine the carousel items size, and I don't think this is the best way to call this callback. If we have `currentIndex` available, are we able to call the callback in the widget layer? I think PageView can be a good example showing how we want to add the current index and the onIndexChange:)